src/main/java/com/thebois/models/beings/Pawn.java

```diff
@@ -48,4 +59,63 @@ private void setRandomDestination() {
         setPath(newPath);
     }
 
+    protected void updateClosestStructure() {
+        if (closestStructure.isPresent() && getFinalDestination().isPresent()) {
+            // If the structure is completed or destination is unreachable pick a new one
+            if (closestStructure.get().isCompleted() || closestStructure.get().getPosition().equals(
+                getFinalDestination().get())) {
+                closestStructure = findIncompleteStructure();
+                setClosestStructureAsFinalDestination();
+            }
+        }
+        else {
+            closestStructure = findIncompleteStructure();
+            setClosestStructureAsFinalDestination();
+        }
+    }
+
+    protected Optional<IStructure> findIncompleteStructure() {
+        final Collection<Optional<IStructure>> structures =
+            getFinder().findNearestStructures(this.getPosition(), 10);
+
+        for (final Optional<IStructure> structure : structures) {
+            if (structure.isPresent()) {
+                if (!structure.get().isCompleted()) {
+                    return structure;
+                }
+            }
+            else {
+                return Optional.empty();
+            }
+        }
+        return Optional.empty();
+    }
```

Comment:
This is the method that should be in the finder itself. You shouldn't ask the finder to get all nearby structures and then find the incomplete one yourself, you should ask the finder for the nearest incomplete structure directly, since that is the responsibility of the finder.